src/Common/Identifiers.hs

```diff
@@ -1,40 +1,96 @@
 {-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE OverloadedStrings #-}
 {- | The types of identifiers used across the compiler.
 
-These are defined as newtypes (rather than as type aliases) so that they cannot
-be accidentally compared with one another.
+The types 'VarId', 'TVarId', 'DConId', and 'TConId' are used in the IR for data
+variables, type variables, data constructors, and type constructors, which each
+inhabit a separate namespace. The underlying 'Identifier' type is used in the
+AST, before these different kinds of identifiers are differentiated by the
+lowering phase.
+
+These are defined as newtypes (rather than as type aliases) that ultimately
+abstract over the underlying 'String' Identifier. We do this for a few reasons:
+
+    * The particular newtype carries semantic meaning in the IR, e.g., 'VarId'
+      is an identifier for data variables and nothing else. This prevents
+      semantically distinct items from being used in place of one another
+      (without explicit coercion). For example, this will prevent accidentally
+      using a type variable identifier as a data constructor.
+
+    * In the future, this may be extended to carry other (non-semantic)
+      metadata, such as source code location.
+
+    * It allows us to freely attach typeclass instances to this type without
+      FlexibleInstances (since 'String' is a type synonym for @[Char]@).
+
+Users should never need to specifically use the data constructor for each
+newtype, e.g., @VarId (Identifier "foo")@. Instead, each identifier belongs to
+the 'IsString' typeclass, that can be written as @fromString "foo"@, so that the
+appropriate type can be inferred from the context where the identifier is used.
```

Comment:
Neat!

Reply:
I believe the `OverloadedStrings` language-extension can help you here :) That way you can in most cases drop the `fromString` call yourself.

Normally, a string literal `"x"` gets the inferred type `String`, but with the `OverloadedStrings` extension enabled the inferred type is instead `isString a => a`. GHC will try to solve the constraint for you :)

Reply:
I already use that in a few modules!